std/digest/xxh.d

```diff
@@ -0,0 +1,3177 @@
+/**
+ * Computes xxHash hashes of arbitrary data. xxHash hashes are either uint32_t,
+ * uint64_t or uint128_t quantities that are like a
+ * checksum or CRC, but are more robust and very performant.
+ *
+$(SCRIPT inhibitQuickIndex = 1;)
+
+$(DIVC quickindex,
+$(BOOKTABLE ,
+$(TR $(TH Category) $(TH Functions)
+)
+$(TR $(TDNW Template API) $(TD $(MYREF XXHTemplate)
+)
+)
+$(TR $(TDNW OOP API) $(TD $(MYREF XXH32Digest))
+)
+$(TR $(TDNW Helpers) $(TD $(MYREF xxh32Of))
+)
+)
+)
+
+ * This module conforms to the APIs defined in `std.digest`. To understand the
+ * differences between the template and the OOP API, see $(MREF std, digest).
+ *
+ * This module publicly imports $(MREF std, digest) and can be used as a stand-alone
+ * module.
+ *
+ * License:   $(HTTP www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+ *
+ * CTFE:
+ * Digests do not work in CTFE
+ *
+ * Authors:
+ * Carsten Schlote, Piotr Szturmaj, Kai Nacke, Johannes Pfau $(BR)
+ * The routines and algorithms are provided by the xxhash.[ch] source
+ * provided at $(I git@github.com:Cyan4973/xxHash.git).
+ *
+ * References:
+ *      $(LINK2 https://github.com/Cyan4973/xxHash, GitHub website of project)
+ *
+ * Source: $(PHOBOSSRC std/digest/xxh.d)
+ *
+ */
+
+/* xxh.d - A wrapper for the original C implementation */
+module std.digest.xxh;
+
+version (X86)
+    version = HaveUnalignedLoads;
+else version (X86_64)
+    version = HaveUnalignedLoads;
+
+///
+@safe unittest
+{
+    //Template API
+    import std.digest.md;
+
+    //Feeding data
+    ubyte[1024] data;
+    XXH_32 xxh;
+    xxh.start();
+    xxh.put(data[]);
+    xxh.start(); //Start again
+    xxh.put(data[]);
+    auto hash = xxh.finish();
+}
+
+///
+@safe unittest
+{
+    //OOP API
+    import std.digest.md;
+
+    auto xxh = new XXH32Digest();
+    ubyte[] hash = xxh.digest("abc");
+    assert(toHexString(hash) == "32D153FF", "Got " ~ toHexString(hash));
+
+    //Feeding data
+    ubyte[1024] data;
+    xxh.put(data[]);
+    xxh.reset(); //Start again
+    xxh.put(data[]);
+    hash = xxh.finish();
+}
+
+public import std.digest;
+
+/* --- Port of C sources (release 0.8.1) to D language below ---------------- */
+
+enum XXH_NO_STREAM = false;
+enum XXH_SIZE_OPT = 0;
+enum XXH_FORCE_ALIGN_CHECK = true;
+enum XXH32_ENDJMP = false;
+
+private import core.bitop : rol, bswap;
+
+/* *************************************
+*  Misc
+***************************************/
+
+alias XXH32_hash_t = uint;
+alias XXH64_hash_t = ulong;
+struct XXH128_hash_t
+{
+    XXH64_hash_t low64; /** `value & 0xFFFFFFFFFFFFFFFF` */
+    XXH64_hash_t high64; /** `value >> 64` */
+}
+
+alias XXH64_canonical_t = ubyte[XXH64_hash_t.sizeof];
+alias XXH128_canonical_t = ubyte[XXH128_hash_t.sizeof];
```

Comment:
XXH64_hash_t is ulong. XXH128_hash_t should be either ucent/Int128 or uint128_t or the emulated structure with 2xulong. So there shouldn't be any padding problem.

I also strongly recommend NOT to use magic numbers. So instead of 64 / 8, we should always use ulong.sizeof /ucent.sizeof  (size in bytes).

I tried to use the class Int128, but this lacks abilties with casting, which is possible with normal integral interger types. I'd like to recommend to rediscuss the issue of makeing (u)cent deprecated. Sooner or later there will be real hardware (e.g. RISC/V) which has native 128bit mnemonics. C provides uint128_t for this. It is either emulated with 2 uint64_t or directly mapped to assembly mnemonics later. D should be prepared for this and should have a native 128bit type as well (and behaving the same way as all integral integer types).

Reply:
There is work in progress to recognise core.int128 as native at a codegen level, so it's unlikely that cent/ucent deprecation will roll back, other than be introduced as an alias to the object module once the deprecation is over.

Reply:
Added extra check for matching structure sizes.

Reply:
> Added extra check for matching structure sizes.

Yeah, that can be a solution. Thanks :)